File: benchmarks/benchmark_cnn.jl
"""
CNN Performance Benchmark (Julia)
Equivalent to benchmark_cnn.py using Lux, Enzyme, and Reactant on the MLX backend.

Usage:
  julia +1.11 benchmarks/benchmark_cnn.jl
"""

using Libdl, Reactant, Lux, Enzyme, Random, Statistics, Optimisers

# ── MLX Plugin Setup ──────────────────────────────────────────────────────────

function load_mlx_plugin!()
    # Pre-load MLX shared library
    mlx_lib = "/Users/thomas/miniforge3/envs/jax/lib/python3.13/site-packages/mlx/lib/libmlx.dylib"
    Libdl.dlopen(mlx_lib, Libdl.RTLD_NOW | Libdl.RTLD_GLOBAL)

    # Find plugin dylib
    plugin_path = joinpath(@__DIR__, "..", "src", "jax_mlx", "libmlx_pjrt_plugin.dylib")
    if !isfile(plugin_path)
        plugin_path = joinpath(@__DIR__, "..", "build", "libmlx_pjrt_plugin.dylib")
    end
    @assert isfile(plugin_path) "Cannot find libmlx_pjrt_plugin.dylib at $plugin_path"

    # Register plugin with Reactant
    mlx_client_ptr = Reactant.XLA.PJRT.MakeClientUsingPluginAPI(plugin_path, "mlx", "MLX")
    mlx_client = Reactant.XLA.PJRT.Client(mlx_client_ptr)
    Reactant.XLA.global_backend_state.clients["mlx"] = mlx_client
    Reactant.XLA.set_default_backend(mlx_client)
    println("MLX plugin loaded ✓")
end

# ── CNN Model ─────────────────────────────────────────────────────────────────

function SimpleCNN()
    return Chain(
        # Block 1 (SAME padding, matching all platforms)
        # cross_correlation=true matches JAX/PyTorch semantics and avoids
        # unnecessary stablehlo.reverse ops in Enzyme's backward pass
        Conv((3, 3), 3 => 64; pad=SamePad(), cross_correlation=true),
        relu,
        Conv((3, 3), 64 => 64; pad=SamePad(), cross_correlation=true),
        relu,
        MaxPool((2, 2)),
        # Block 2
        Conv((3, 3), 64 => 128; pad=SamePad(), cross_correlation=true),
        relu,
        Conv((3, 3), 128 => 128; pad=SamePad(), cross_correlation=true),
        relu,
        MaxPool((2, 2)),
        # Block 3
        Conv((3, 3), 128 => 256; pad=SamePad(), cross_correlation=true),
        relu,
        # Global average pooling → Dense head
        GlobalMeanPool(),
        FlattenLayer(),
        Dense(256, 128, relu),
        Dense(128, 10),
    )
end

# ── Loss Function ─────────────────────────────────────────────────────────────

function cross_entropy_loss(model, ps, st, x, y)
    logits, st_ = model(x, ps, st)
    # logits is (10, batch) in Lux's format
    # log-softmax along dim 1 (classes)
    log_probs = logits .- Lux.logsumexp(logits; dims=1)
    # One-hot via broadcasting: (10,1) .== (1,batch) → (10, batch)
    classes = reshape(1:10, 10, 1)
    labels = reshape(y, 1, :)
    one_hot = Float32.(classes .== labels)
    return -sum(one_hot .* log_probs) / Float32(size(logits, 2))
end

# ── Train Step (Reactant-compatible) ──────────────────────────────────────────

function train_step(model, ps, st, x, y, lr)
    # Use Enzyme for gradients through Reactant tracing
    dps = Enzyme.make_zero(ps)
    _, loss = Enzyme.autodiff(
        Enzyme.ReverseWithPrimal,
        cross_entropy_loss,
        Active,
        Const(model),
        Duplicated(ps, dps),
        Const(st),
        Const(x),
        Const(y),
    )
    # SGD update: ps = ps - lr * dps
    ps_new = Lux.recursive_map((p, g) -> p .- lr .* g, ps, dps)
    return ps_new, loss
end

# ── Benchmark ─────────────────────────────────────────────────────────────────

function benchmark_cnn(;
    batch_size::Int=128,
    image_size::Int=64,
    num_warmup::Int=30,
    num_runs::Int=20,
    lr::Float32=0.01f0,
)
    println("CNN Performance Benchmark (Julia/Lux/Reactant)")
    println("Config: batch_size=$batch_size, image_size=$(image_size)x$(image_size)x3")
    println("Task: Forward + Backward pass with SGD\n")

    # Load plugin
    load_mlx_plugin!()

    # Build model
    model = SimpleCNN()
    rng = Random.Xoshiro(0)
    ps, st = Lux.setup(rng, model)
    ps = ps |> Lux.f32

    # Create random data (Lux uses WHCN format: Width x Height x Channels x Batch)
    x = randn(Float32, image_size, image_size, 3, batch_size)
    y = rand(1:10, batch_size)

    # Convert to Reactant arrays
    x_r = Reactant.to_rarray(x)
    y_r = Reactant.to_rarray(y)
    ps_r = Reactant.to_rarray(ps)
    st_r = Reactant.to_rarray(st)

    # Compile the train step
    println("Compiling train step...")
    t_compile_start = time()
    compiled_step = @compile train_step(model, ps_r, st_r, x_r, y_r, lr)
    t_compile = time() - t_compile_start
    println("Compilation done in $(round(t_compile; digits=1))s ✓\n")

    # Warmup
    println("Warming up ($num_warmup steps)...")
    for i in 1:num_warmup
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
    end
    println("Warmup done ✓\n")

    # Timed runs
    println("="^50)
    println("Timed runs")
    println("="^50)
    times = Float64[]
    for i in 1:num_runs
        t0 = time_ns()
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
        loss_val = Float64(loss)  # forces materialization (block until ready)
        t1 = time_ns()
        dt_ms = (t1 - t0) / 1e6
        push!(times, dt_ms)
        println("  Run $i: $(round(dt_ms; digits=2))ms (loss: $(round(loss_val; digits=4)))")
    end

    # Exclude first run (may include deferred compilation)
    steady = times[2:end]
    mean_t = mean(steady)
    std_t = std(steady)
    min_t = minimum(steady)
    println("\n  Mean: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms  Min: $(round(min_t; digits=2))ms  (excl. first run)")

    println("\n" * "="^50)
    println("SUMMARY")
    println("="^50)
    println("Julia/Lux/Reactant/MLX: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms")
end

# ── Main ──────────────────────────────────────────────────────────────────────

benchmark_cnn()

File: benchmarks/benchmark_cnn.py
"""
CNN Training Benchmark: JAX-MLX
Full training loop with forward + backward + SGD.

Usage:
  python benchmarks/benchmark_cnn.py
"""

import os
if "MLX_PJRT_DEBUG" in os.environ:
    del os.environ["MLX_PJRT_DEBUG"]
if "JAX_PLATFORMS" not in os.environ:
    os.environ["JAX_PLATFORMS"] = "mlx"

import jax
import jax.numpy as jnp
from flax import linen as nn
import time
import numpy as np


class SimpleCNN(nn.Module):
    """Deep CNN for benchmarking."""

    @nn.compact
    def __call__(self, x):
        # Block 1
        x = nn.Conv(features=64, kernel_size=(3, 3), padding='SAME')(x)
        x = nn.relu(x)
        x = nn.Conv(features=64, kernel_size=(3, 3), padding='SAME')(x)
        x = nn.relu(x)
        x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2))

        # Block 2
        x = nn.Conv(features=128, kernel_size=(3, 3), padding='SAME')(x)
        x = nn.relu(x)
        x = nn.Conv(features=128, kernel_size=(3, 3), padding='SAME')(x)
        x = nn.relu(x)
        x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2))

        # Block 3
        x = nn.Conv(features=256, kernel_size=(3, 3), padding='SAME')(x)
        x = nn.relu(x)

        x = jnp.mean(x, axis=(1, 2))  # Global average pooling
        x = nn.Dense(features=128)(x)
        x = nn.relu(x)
        x = nn.Dense(features=10)(x)
        return x


def loss_fn(params, model, x, y):
    """Cross-entropy loss."""
    logits = model.apply(params, x)
    one_hot = jax.nn.one_hot(y, 10)
    return -jnp.mean(jnp.sum(jax.nn.log_softmax(logits) * one_hot, axis=-1))


def create_train_step(model, lr=0.01):
    @jax.jit
    def train_step(params, x, y):
        loss, grads = jax.value_and_grad(lambda p: loss_fn(p, model, x, y))(params)
        params = jax.tree.map(lambda p, g: p - lr * g, params, grads)
        return params, loss
    return train_step


def benchmark_mlx(batch_size=128, image_size=64, num_warmup=30, num_runs=20):
    """Benchmark CNN training on MLX."""
    try:
        device = jax.devices("mlx")[0]
    except Exception as e:
        print(f"MLX device not available: {e}")
        return None, None

    model = SimpleCNN()
    input_shape = (batch_size, image_size, image_size, 3)

    with jax.default_device(device):
        rng = jax.random.PRNGKey(0)
        params = model.init(rng, jnp.ones(input_shape))
        key = jax.random.PRNGKey(42)
        x = jax.random.normal(key, input_shape)
        y = jax.random.randint(key, (batch_size,), 0, 10)
        train_step = create_train_step(model)

        # Warmup
        print(f"Warming up ({num_warmup} steps)...")
        for _ in range(num_warmup):
            params, loss = train_step(params, x, y)
            loss.block_until_ready()

        # Timed runs
        print(f"\n{'='*50}")
        print("Timed runs")
        print(f"{'='*50}")
        times = []
        for i in range(num_runs):
            start = time.perf_counter()
            params, loss = train_step(params, x, y)
            loss.block_until_ready()
            end = time.perf_counter()
            times.append(end - start)
            print(f"  Run {i+1}: {times[-1]*1000:.2f}ms (loss: {float(loss):.4f})")

    mean_time = np.mean(times) * 1000
    std_time = np.std(times) * 1000
    min_time = np.min(times) * 1000
    print(f"\n  Mean: {mean_time:.2f}ms ± {std_time:.2f}ms  Min: {min_time:.2f}ms")
    return mean_time, std_time


def main():
    batch_size = 128
    image_size = 64

    print("CNN Training Benchmark (JAX-MLX)")
    print(f"Config: batch_size={batch_size}, image_size={image_size}x{image_size}x3")
    print(f"Model: Conv(64)×2 → Pool → Conv(128)×2 → Pool → Conv(256) → GAP → Dense(128) → Dense(10)")
    print(f"Task: Forward + Backward pass with SGD\n")

    mean, std = benchmark_mlx(batch_size, image_size)

    if mean is not None:
        print(f"\n{'='*50}")
        print("SUMMARY")
        print(f"{'='*50}")
        print(f"JAX-MLX: {mean:.2f}ms ± {std:.2f}ms")


if __name__ == "__main__":
    main()
